main/src/library/Fixpoint3/Phase/Compiler.flix

```diff
@@ -0,0 +1,584 @@
+    use Fixpoint3.Predicate.PredType.{Full, Delta, New}
+
+    ///
+    /// Compile the given Datalog program `d` to RAM.
+    /// First insert all facts in `d` into the set of facts.
+    /// Second compile all rules in `d` according to the given stratification `strat`.
+    /// This means to compile all rules in stratum 0, then stratum 1 and so on.
+    /// A rule belongs to the same stratum as its head predicate.
+    ///
+    @Internal
+    pub def compile(d: Datalog, db: Map[RelSym, Map[Vector[Boxed], Boxed]], strat: Map[PredSym, (Int32, Int32)]): RamProgram = match d {
+        case Datalog(facts, rules) => region rc {
+            let predicates = Predicate.initialize(d, db);
+            let stmts = MutList.empty(rc);
+            let compiledFacts = compileFacts(facts, db, predicates);
+            // Create map of stratum to list of rules in stratum
+            let counter = Counter.fresh(rc);
+            let pseudoStratums = MutMap.empty(rc);
+            rules |> Vector.forEach(match rule -> match rule {
+                case Constraint(HeadAtom(p, _, _), _) =>
+                    let (pseudoStratum, stratum) = Map.getWithDefault(p, (0, 0), strat);
+                    let pseudoMap = MutMap.getOrElsePut(pseudoStratum, MutMap.empty(rc), pseudoStratums);
+                    MutMap.putWith(List.append, stratum, rule :: Nil, pseudoMap)
+            });
+            pseudoStratums |> MutMap.forEach(_ -> pseudStrat -> {
+                let runInParallel = pseudStrat |> MutMap.foldLeft(acc -> stratRules -> {
+                    let stmtsForPseudo = MutList.empty(rc);
+                    compileStratum(stmtsForPseudo, stratRules, predicates, counter);
+                    (MutList.toVector(stmtsForPseudo) |> RamStmt.Seq) :: acc
+                }, Nil);
+                if(List.size(runInParallel) > 1) {
+                    let codeBlock = List.toVector(runInParallel) |> RamStmt.Par;
+                    MutList.push(codeBlock, stmts)
+                } else match List.head(runInParallel) {
+                    case Some(v) => MutList.push(v, stmts)
+                    case None => ()
+                }
+            });
+            let stmt = MutList.toVector(stmts) |> RamStmt.Seq;
+            RamProgram.Program(stmt, compiledFacts, predicates, (Map#{}, Map#{}))
+        }
+        case _ => bug!("Datalog Boxing bug")
+    }
+
+    ///
+    /// Collect all rules in `d` and assign them a number, deterministically.
+    ///
+    @Internal
+    pub def enumerateRules(d: Datalog): Vector[(Int32, Constraint)] = match d {
+        case Datalog(_, rules) => Vector.mapWithIndex(i -> r -> (i, r), rules)
+        case _ => unreachable!()
+    }
+
+    ///
+    /// Functional relations should be treated as `Relational`, but they do not posses any `PredSym`, indicated
+    /// by assigning them -1.
+    ///
+    @Internal
+    pub def functionalRelSym(arity: Int32): RelSym = RelSym.Symbol(PredSym.PredSym("", -1i64), arity, Relational)
```

Comment:
This looks a bit scary to me, can you add explanation why its okay for all of these to reuse the same "-1" sym